src/plumpy/process_states.py

```diff
@@ -225,7 +227,7 @@ async def execute(self) -> State:  # type: ignore # pylint: disable=invalid-over
             try:
                 try:
                     self._running = True
-                    result = self.run_fn(*self.args, **self.kwargs)
+                    result = await self.run_fn(*self.args, **self.kwargs)
```

Comment:
Maybe the better way to do this is using https://docs.python.org/3/library/inspect.html#inspect.isawaitable
🤔 (see e.g. https://textual.textualize.io/blog/2023/03/15/no-async-async-with-python/)

Reply:
What is mentioned in the blog post that @chrisjsewell mentioned in https://github.com/aiidateam/plumpy/pull/272/files#r1257558025 with `await_me_maybe` would be the better design IMO. I think but we already do it the same way in the `Process` class so this is at least consistent. Maybe we can instead improve the doc a bit on the places where we use `ensure_coroutine`. Something like 
```
We wrap `run_fn` to a coroutine so we can apply await on it, even it if it was not a coroutine in the first place. This allows the same usage of async and non-async function with the await syntax while not changing the program logic.
```
At least this is my understanding why we do this. (Also would add something like this to _run_task in class Process)

Reply:
Thanks @agoscinski , I've added your comment.